File: Task/Decorate-sort-undecorate-idiom/PascalABC.NET/decorate-sort-undecorate-idiom.pas
##
function Schwartzian<T, TKey>(source: sequence of T; decorator: T-> TKey) :=
   source.Select(x -> (x, decorator(x)))
   .orderby(x -> x[1])
   .select(x -> x[0]);

var test: sequence of string := |'Rosetta', 'Code', 'is', 'a', 'programming', 'chrestomathy', 'site'|;
Schwartzian(test, x -> x.length).Println;

File: Task/Decorate-sort-undecorate-idiom/C-sharp/decorate-sort-undecorate-idiom.cs
public static IEnumerable<T> Schwartzian1<T, TKey>(IEnumerable<T> source, Func<T, TKey> decorator) =>
    source.Select(item => (item, key: decorator(item)))
        .OrderBy(tuple => tuple.key)
        .Select(tuple => tuple.item);

public static IEnumerable<T> Schwartzian2<T, TKey>(IEnumerable<T> source, Func<T, TKey> decorator) =>
    from item in source
    let key = decorator(item)
    orderby key
    select item;

//Call:
string[] array = {"Rosetta", "Code", "is", "a", "programming", "chrestomathy", "site"};
Console.WriteLine(string.Join(" ", Schwartzian1(array, i => i.Length)));
